### DISCUSSION TITLE
rate sensor array entity.attributes.rates.  Why only 46 rates (midnight to 21:30) and what range is average rate taken from?

### DISCUSSION THREAD

**Discussion by deltamelter** [BottlecapDave/HomeAssistant-OctopusEnergy#283] [Q&A]
Knocked up an Apex-Chart but struggling with the average.  The current_day_average_rate seems to be based on midnight to 21:30 and I wanted for the whole day, so wrote a transform, but the sensor's rate array only contains 45 items too gives me a different average. :rofl:    
![image](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/assets/27061628/c82bc190-d40e-41dc-a3fe-43afc6a6a9f8)
How is the current day average calculated, and would it be possible to include full 24 hours in the rates array?  ie from 21:30 to 21:30?

**Comment by BottlecapDave** [comment]
HI and sorry that you're having issues. The rates are retrieved every 30 minutes for today and tomorrow and are broken down into 30 minute chunks on the hour and half hour. The min, max and average rate is calculated for the current day. All dates/times in the rates are stored in UTC format, so will be from 23:00 to 23:00 during BST. The graph you have shown looks like this is the case (ending at 23:00). Because of all these factors, i'd expect 96 entries. If you have less than this, then it might be due to the data available from OE. 

I assume you're on agile based on your graph? If so, which specific tariff code (located as an attribute in the current rate sensor). I believe this data is only available from ~4pm.

Also could you post a data reported by the `rates` attribute?

**Comment by deltamelter** [comment]
Hi, 
It looks like after 4pm (when the next day is published) and up until midnight there are 94 items in rates, but after midnight and until next 4pm update, there are only 46.  It starts with 2300 previous day and ends with 21:30 (-22:00), but as we roll into next day (or some time before maybe) this gets truncated back to just 46 entries from 23:00 to 21:30
The graph runs to item starting 21:30 (22:30 BST)
Tariff: E-1R-AGILE-FLEX-22-11-25-H

![image](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/assets/27061628/b77eb2e5-2b39-40f3-ba9f-de7dda588323)

![image](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/assets/27061628/1570ec23-fbd4-4e6c-95c7-b19a99421936)

**Reply by BottlecapDave** [reply]
Yeah so as mentioned, this is because OE only post a little over a day's worth of results. When you move into the next day you're still in the current agile period, but before the next period has been published which is why you have 1 hour shy of a full day's worth of data (24 hours in 30 min chunks). The data cuts down when you move into the next day as the dates of data it's looking for has shifted.

I've manually looked up the rates until 30th June and I only get data up to 22:00 UTC of the 28th June. So unfortunately this is working as intended as this is the maximum data that OE provide.

**Reply by deltamelter** [reply]
So do to create the "Current day average rate" from 24 hours of rates while there are 47 hours worth of data, or is it recalculated every time the rates are retrieved from OE?  What times are you taking the average from?  midnight UTC to midnight UTC? ("from")  
 
I took the average shown in the graph from last entry, 21:30 and the 24 hour before it, and this seemed to me to better reflect the upcoming days average rate.   After midnight, my calculated average is now only from 23hours and less reflective. 

OE only publishes up to 21:30 (to 22:00) each day, yes, but they also return the previous days, I'm not asking for what isn't there yet, but I am asking to go backwards from the last entry and provide at least 24 hours.

**Reply by deltamelter** [reply]
OK, I saw from observation that the "Current day average rate" is recalculated from what is the for "the day" so after 4pm (OE publish next day) the average is of the last 48 half hours , I calculated exactly the same in sheets 
2023-06-27T00:00:00Z | 2023-06-27T00:30:00Z
to
2023-06-27T23:30:00Z | 2023-06-28T00:00:00Z
gave average of 21.81
`Current day min rate
15.88
Current day max rate
34.23
Current day average rate
21.81`

Then after midnight the average seems to only base on data for a couple of records and for the next hour until 1am bst  
`Current day min rate
15.88
Current day max rate
16.76
Current day average rate
16.32` 

After 1am:
`Current day min rate
15.96
Current day max rate
34.23
Current day average rate
23.12`

Attaching the attributes for the entities from both before and after 1am
[rates-24.txt](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/files/11892407/rates-24.txt)

**Reply by BottlecapDave** [reply]
Yup looks like a couple of BST bugs causing some miscalculations. It should be midnight to midnight local time (BST/GMT). I'll look at fixing this soon.

Returning previous day rates in the data would be a breaking change as other people might be expecting the current data.

**Reply by BottlecapDave** [reply]
The BST bugs should be fixed in the latets release v7.4.4

**Comment by deltamelter** [comment]
Attaching the data as Attributes from entity, (from both before 4pm and after in same file) which might be a bit more readable :grin:
[rates.txt](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/files/11886087/rates.txt)
